pyupgrade.py

```diff
@@ -1537,6 +1535,37 @@ def generic_visit(self, node):  # type: (ast.AST) -> None
         super(FindPy3Plus, self).generic_visit(node)
 
 
+def reversed_walk(node):  # type: (ast.AST) -> Generator[ast.AST, None, None]
+    todo = [node]
+    while todo:
+        node = todo.pop()
+        todo.extend(ast.iter_child_nodes(node))
+        yield node
+
+
+def targets_referenced_after_node(root_node, node, targets):
+    # type: (ast.AST, ast.AST, Union[ast.expr, ast.Tuple]) -> bool
+    """
+    Search inside the given root node whether the targets have been referenced
+    after the given node. Walks from behind until the node have been reached.
+    """
+    for x in reversed_walk(root_node):
```

Comment:
alternatively you forward walk (don't need the other function) and then you start looking when the node is encountered

Reply:
I came up with:
```python
todo = collections.deqeu(func_node.body)
while todo:
    node = todo.popleft()
    ....
```
What do you think?

Reply:
I'd prefer trying to make it work with `ast.walk` if possible -- failing that whatever works

Reply:
Can't figure out how to achieve the same with `ast.walk`

```python
    visited = False
    todo = collections.deque(func_node.body)

    while todo:
        node = todo.popleft()

        if node == forloop_node:
            visited = True
            continue

        if not visited:
            continue

        if (
            isinstance(node, ast.Name) and
            isinstance(node.ctx, ast.Load) and
            node.id in names
        ):
            return True

        todo.extend(ast.iter_child_nodes(node))

    return False
```

How to prevent walking inside the for loop? Have any ideas?

Reply:
Dammit, I'm too bad at walking trees 😒

Reply:
adjust the `visited` assignment until the last element of the `yield` expression maybe

Reply:
Too complicated, maybe node visitor would help? I'll try that.

Reply:
if you're using a node visitor you could reuse the existing walk and remove the loops that are referenced after -- this would also keep it 1-pass :) (you might need to store some additional state, but you could use `visit_Name` directly)

Reply:
Strange just happened. I wrote the visitor and tests passed :D And it's less code.

Reply:
Just a thought, can we skip the nodes till for loop depend on offset information?

Reply:
not always, for instance

```python
f'''
{foo}
'''
```

reports an offset of -1 on python<3.8

Reply:
Please review in 81dc036f6a52e32a3dd1e992334c24fc09cea12b  how it's done via visitor and stack depth level checking.

Reply:
cool I'll poke at this, might be another way to approach it I'll see what I can do 👍

Reply:
Thank you so much!

Reply:
> adjust the `visited` assignment until the last element of the `yield` expression maybe

FTR, looks like it's impossible since `ast.walk` does undefined-order traversal.